### DISCUSSION TITLE
What is the maximum policies or SAs supported by strongswan?

### DISCUSSION THREAD

**Discussion by mavis-yang** [strongswan/strongswan#477] [General]
Hi expert,

Hoping you are doing well :) 

I would like to check what's the maximum policies or SAs supported by strongswan.

Is there MACROs defined by strongswan to limit the policy or SA number?
-- if yes, could you share where it is defined in the source code.

Or the number of policy or SA depends on Linux capacity, e.g. memory?
-- If yes, could you share where Linux defines the number?

And whether the number is impacted by the number of route in Linux?

Looking forward your response coming soon.

Thanks.

Regards,
Mavis

**Comment by tobiasbrunner** [comment]
> Is there MACROs defined by strongswan to limit the policy or SA number?

No, there are no hard limits on either. If you consider lots of IKE_SAs, have a look at [this wiki page](https://wiki.strongswan.org/projects/strongswan/wiki/IkeSaTable) (and maybe [this one](https://wiki.strongswan.org/projects/strongswan/wiki/JobPriority) too).

> Or the number of policy or SA depends on Linux capacity, e.g. memory?

Sure, the memory obviously limits this at some point (but I doubt you will ever run into this issue) and policy lookups will be slower if there are a lot (recent kernels should be better, though, and the `spdh_thresh` options might help in some cases).

> And whether the number is impacted by the number of route in Linux?

Maybe at some point, but you might not need individual routes (really depends on the scenario) and could just disable route installation by strongSwan and maybe use a single route for all possible peers.

**Reply by mavis-yang** [reply]
Tobias,
Thanks for your info.

If there is no bottleneck for SA number. How to make huge number of SAs' establishment quickly?

The latest testing, when there is **3000** connections (3000 IKE SA + 3000 IPsec SA), it spends about **40min** for all SA creations.
-- test peer is with auto=start, remote peer is with auto=add
-- Test time is from strongswan service start (ipsec start; auto=start) to there is 3000 IPsec SA
-- strongswan v5.9.2
-- no big gap between starter (ipsec.conf) and swanctl (swanctl.conf)
-- when the conn number increase, time of single connection estimation becomes slower.

The time seems too long to meet customer.
Is there solution to make SA creation concurrently to shorten the total time? 

Regards,
Yang Li

**Reply by tobiasbrunner** [reply]
Please refer to the documents I linked before.

**Reply by mavis-yang** [reply]
Hi Tobias,

From you have linked doc, my understanding is
* [this wiki page](https://wiki.strongswan.org/projects/strongswan/wiki/IkeSaTable): It is optimized by ikesa_table_size, which makes SA look up quickly.
* [this one](https://wiki.strongswan.org/projects/strongswan/wiki/JobPriority): It is to adjust task priority. Does it make SA establishment concurrently?

The expectation is to shorten SA establishment time from 40min to 2min as initiator.

Regards,
Yang Li

**Reply by tobiasbrunner** [reply]
> The expectation is to shorten SA establishment time from 40min to 2min as initiator.

As initiator?  Why would a single initiator need to create thousands of SAs? Note that if you want to test lots of initiations against a gateway, you might want to look at the [load-tester plugin](https://wiki.strongswan.org/projects/strongswan/wiki/LoadTests).

**Reply by mavis-yang** [reply]
>  Why would a single initiator need to create thousands of SAs? 

It is deployed in one CU network element, which act as Initiator to connect neighbor CUs for Xn traffic.

**Reply by tobiasbrunner** [reply]
> It is deployed in one CU network element, which act as Initiator to connect neighbor CUs for Xn traffic.

Sorry, no idea what that means.

**Reply by Thermi** [reply]
Configuration lookup is in O(n), so it's not advisable to have a configuration for each other site. It's better to instanciate new connections by providing command line parameters for source and destination as made possible by the DMVPN patches by Timo Teras.

Regarding instanciating lots of "simple" load testing connections, try to make use of load-tester (part of strongSwan). It's made for the specific purpose of negotiating lots of IKE_SAs and CHILD_SAs.

To enable concurrent working on the IKE_SA table, configure IKE_SA table segmentation by making it a hash table with lots of segments. You do that using [the instructions for it on the wiki](https://wiki.strongswan.org/projects/strongswan/wiki/IkeSaTable).
